### PR TITLE
Add support for Python 3.12

### PR DESCRIPTION
**Pull request by nanonyme**
Continuation of https://github.com/apache/buildstream/pull/1872

### CODE REVIEW

File: .github/common.env [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Shared common variables
 
-CI_IMAGE_VERSION=master-862722333
+CI_IMAGE_VERSION=master-1091773482
 CI_TOXENV_MAIN=py37,py38,py39,py310,py311
 CI_TOXENV_PLUGINS=py37-plugins,py38-plugins,py39-plugins,py310-plugins,py311-plugins
```
Comment (abderrahim) [inline .github/common.env]: I think you need to add py312 to the above two lines to make it run in CI (I wonder why we have this)

Comment (nanonyme) [reply .github/common.env]: I assume this is so we can test on same Fedora version with multiple Python versions. But if so, no idea why so many Fedora versions are used....

Comment (nanonyme) [reply .github/common.env]: Fixed.

**Comment by abderrahim** [comment]
So I tried to run the tests locally, and it seems we can't easily get 3.7 and 3.12 at the same time. 3.12 doesn't work with the current dependencies and trying to update them seems to break 3.7.

**Comment by juergbi** [comment]
> So I tried to run the tests locally, and it seems we can't easily get 3.7 and 3.12 at the same time. 3.12 doesn't work with the current dependencies and trying to update them seems to break 3.7.

I suppose that's not surprising given that Python 3.7 reached end of life 5 months ago. I think it's time to drop Python 3.7 support in BuildStream master as well.

**Comment by nanonyme** [comment]
@juergbi who needs to be present for that decision?

**Comment by nanonyme** [comment]
@abderrahim both pyroaring and ruamel.yaml.clib *claim* to support Python 3.7 with the new versions that support Python 3.12.  Was the problematic dependency grpcio?

**Comment by nanonyme** [comment]
@juergbi can I have another pipeline? It would be useful to be able to run pipelines without approval every time.

**Comment by nanonyme** [comment]
Looks like grpcio 1.59.0 or newer is required for Python 3.12 support. These should support Python 3.12.

**Comment by nanonyme** [comment]
@abderrahim also how did you handle pluginbase? It looks like there simply is no  released version of it that supports Python 3.12.  There was a deprecation warning about pending removal of asyncore, I guess it's now removed.

**Comment by nanonyme** [comment]
I am seeing if I can surgically create dependencies that work on all Python versions. It would be better to drop Python 3.7 support in a separate PR if possible.

**Comment by nanonyme** [comment]
@juergbi there's some random errors with buildbox-casd. It's problematic that tests do not propagate buildbox-casd logs when they fail.

**Comment by abderrahim** [comment]
> @juergbi there's some random errors with buildbox-casd. It's problematic that tests do not propagate buildbox-casd logs when they fail.

If this is the buildgrid service-tests, you can see that buildgrid is crashing because it's failing to import `redis`

**Comment by nanonyme** [comment]
@abderrahim no, I mean Debian 10, Fedora 37 and Fedora 38. There's comments about buildbox-casd not starting as expected. And we're also seeing the pesky hardlink failures.

```
    [00:00:00] FAILURE [4a47c98a] build-shell/buildtree.bst: Failed to link '/home/testuser/buildstream/.tox/py311/tmp/test_shell_pull_cached_buildtr2/cache/artifacts/refs/test/build-shell-buildtree/4a47c98a10df39e65e99d471f96edc5b58d4ea5b9b1f221d0be832a8124b8099 -> /home/testuser/buildstream/.tox/py311/tmp/test_shell_pull_cached_buildtr2/cache/artifacts/refs/test/build-shell-buildtree/38b730aec720dbddba843c9e3ce1ad8073e39d16cc4d25f752bc89d867fbaec1': [Errno 17] File exists: '/home/testuser/buildstream/.tox/py311/tmp/test_shell_pull_cached_buildtr2/cache/artifacts/refs/test/build-shell-buildtree/4a47c98a10df39e65e99d471f96edc5b58d4ea5b9b1f221d0be832a8124b8099' -> '/home/testuser/buildstream/.tox/py311/tmp/test_shell_pull_cached_buildtr2/cache/artifacts/refs/test/build-shell-buildtree/38b730aec720dbddba843c9e3ce1ad8073e39d16cc4d25f752bc89d867fbaec1'
````

**Comment by nanonyme** [comment]
Beyond these it seems tests now pass.

**Comment by nanonyme** [comment]
Can I have another pipeline? I pushed commit that may stabilize tests.

File: src/buildstream/utils.py [hunk 2]
```diff
@@ -367,6 +367,15 @@ def safe_copy(src: str, dest: str, *, copystat: bool = True, result: Optional[Fi
                 result.failed_attributes.append(dest)
 
 
+def _same_file(src: str, dest: str):
+    src_stat = os.stat(src)
+    dest_stat = os.stat(dest)
+    if not src_stat.st_ino or not dest_stat:
```
Comment (nanonyme) [inline src/buildstream/utils.py]: There is a bug here so PR must not be merged but I'm letting current pipeline finish.

**Comment by nanonyme** [comment]
The commit did not help.
`tests/integration/shellbuildtrees.py::test_shell_pull_cached_buildtree[pull-without-buildtree] FAILED [ 72%]` this is still esoterically failing due to file existing that should not.

**Comment by nanonyme** [comment]
It looks like it's always the same test that fails.

**Comment by abderrahim** [comment]
> It looks like it's always the same test that fails.

Yeah. It looks like some race condition or something. We need to figure out where it is coming from :/

**Comment by abderrahim** [comment]
> > It looks like it's always the same test that fails.
> 
> Yeah. It looks like some race condition or something. We need to figure out where it is coming from :/

Indeed it is. For some reason, bst is launching 2 jobs at once to pull the same element.

**Comment by abderrahim** [comment]
@nanonyme https://github.com/apache/buildstream/pull/1882 fixes the "pesky hardlink failure", can you please rebase now that it's merged.

**Comment by nanonyme** [comment]
@abderrahim pushed, removed the last commit that was irrelevant.

**Comment by nanonyme** [comment]
Only buildgrid test is now failing.

**Comment by abderrahim** [comment]
Yup, and I believe it's the buildgrid image that's broken (and our tests broke because we're using the nightly image of buildgrid rather than a fixed version)

**Review by abderrahim — APPROVED** [review]

**Pull request merged**